drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,827 @@
+		return ret2;
+	}
+	return 0;
+}
+
+static int pinnacle_handle_interrupt(const struct device *dev)
+{
+	const struct pinnacle_config *config = dev->config;
+	struct pinnacle_data *drv_data = dev->data;
+	union pinnacle_sample *sample = &drv_data->sample;
+
+	int rc;
+
+	rc = pinnacle_sample_fetch(dev, sample);
+	if (rc) {
+		LOG_ERR("Failed to read data packets");
+		return rc;
+	}
+
+	if (config->data_mode == PINNACLE_DATA_MODE_ABS) {
+		if (config->clipping_enabled && !pinnacle_is_idle_sample(sample)) {
+			pinnacle_clip_sample(dev, sample);
+			if (config->scaling_enabled) {
+				pinnacle_scale_sample(dev, sample);
+			}
+		}
+
+		input_report_abs(dev, INPUT_ABS_X, sample->abs_x, false, K_FOREVER);
+		input_report_abs(dev, INPUT_ABS_Y, sample->abs_y, false, K_FOREVER);
+		input_report_abs(dev, INPUT_ABS_Z, sample->abs_z, true, K_FOREVER);
+	} else if (config->data_mode == PINNACLE_DATA_MODE_REL) {
+		input_report_rel(dev, INPUT_REL_X, sample->rel_x, false, K_FOREVER);
+		input_report_rel(dev, INPUT_REL_Y, sample->rel_y, true, K_FOREVER);
+	}
+
+	return 0;
+}
+
+static void pinnacle_data_ready_gpio_callback(const struct device *dev, struct gpio_callback *cb,
+					      uint32_t pins)
+{
+	struct pinnacle_data *drv_data = CONTAINER_OF(cb, struct pinnacle_data, dr_cb_data);
+
+	k_work_submit(&drv_data->work);
+}
+
+static void pinnacle_work_cb(struct k_work *work)
+{
+	struct pinnacle_data *drv_data = CONTAINER_OF(work, struct pinnacle_data, work);
+
+	pinnacle_handle_interrupt(drv_data->dev);
+}
+
+int pinnacle_init_interrupt(const struct device *dev)
+{
+	struct pinnacle_data *drv_data = dev->data;
+	const struct pinnacle_config *config = dev->config;
+	const struct gpio_dt_spec *gpio = &config->dr_gpio;
+
+	int rc;
+
+	drv_data->dev = dev;
+	drv_data->work.handler = pinnacle_work_cb;
+
+	/* Configure GPIO pin for HW_DR signal */
+	rc = gpio_is_ready_dt(gpio);
+	if (!rc) {
+		LOG_ERR("GPIO device %s/%d is not ready", gpio->port->name, gpio->pin);
+		return -ENODEV;
+	}
+
+	rc = gpio_pin_configure_dt(gpio, GPIO_INPUT);
+	if (rc) {
+		LOG_ERR("Failed to configure %s/%d as input", gpio->port->name, gpio->pin);
+		return rc;
+	}
+
+	rc = gpio_pin_interrupt_configure_dt(gpio, GPIO_INT_EDGE_TO_ACTIVE);
+	if (rc) {
+		LOG_ERR("Failed to configured interrupt for %s/%d", gpio->port->name, gpio->pin);
+		return rc;
+	}
+
+	gpio_init_callback(&drv_data->dr_cb_data, pinnacle_data_ready_gpio_callback,
+			   BIT(gpio->pin));
+
+	rc = gpio_add_callback(gpio->port, &drv_data->dr_cb_data);
+	if (rc) {
+		LOG_ERR("Failed to configured interrupt for %s/%d", gpio->port->name, gpio->pin);
+		return rc;
+	}
+
+	return 0;
+}
+
+static int pinnacle_init(const struct device *dev)
+{
+	const struct pinnacle_config *config = dev->config;
+	const struct pinnacle_bus *bus = &config->bus;
+
+	int rc;
+	uint8_t value;
+
+	if (!bus->is_ready(bus)) {
+		return -ENODEV;
+	}
+
+	rc = bus->read(bus, PINNACLE_REG_FIRMWARE_ID, &value);
```

Comment:
hey how about adding some wrapper for the bus functions so that these would show as

`rc = pinnacle_read(dev, PINNACLE_REG_FIRMWARE_ID, &value);`

I think the code would look a bit smoother, but I may be overthinking it, up to you :-)

Reply:
Updated, the code looks more cleaner now.